File: src/namagomi/minecraft/api/data/namagomiData.ts
import {GitTree} from "../github/GitTree";
import fs, {createWriteStream} from "fs";
import {pipeline} from "stream/promises";
import fetch from "electron-fetch";
import {namagomiDataFileUrlBase, namagomiFileUrlBase} from "../../../settings/config";
import path from "path";
import {mainDir, namagomiCache} from "../../../settings/localPath";

export interface NamagomiCache {
    data: {
        name: string
        sha: string
    }[]
    mods: string
}

export async function downloadAllDataFiles(branch: string, side: string) {
    const tree = await new GitTree().build('NamagomiNetwork', 'Namagomi-mod', branch);
    const dataSha = tree.getData('data').data.sha;
    const dataTree = await new GitTree().build('NamagomiNetwork', 'Namagomi-mod', dataSha);
    const dataPaths = await dataTree.getAllFilePaths()
    const dataSubDirs = await dataTree.getAllDirectoryPaths()

    if (!fs.existsSync(namagomiCache(side))) createEmptyJson(namagomiCache(side))
    const cacheJson = JSON.parse(fs.readFileSync(namagomiCache(side), 'utf8')) as NamagomiCache

    await Promise.all(dataPaths.map(async cfgPath => {
        if (!fs.existsSync(mainDir(side))) fs.mkdirSync(mainDir(side))
        dataSubDirs.map(async (dir) => {
            const absDir = path.join(mainDir(side), dir)
            if (!fs.existsSync(absDir)) fs.mkdirSync(absDir)
        })

        const sha = (await dataTree.getData(cfgPath)).data.sha

        const findIndex = cacheJson.data.findIndex((d: { name: string, sha: string }) => d.name === cfgPath)

        if (findIndex === -1 || cacheJson.data[findIndex].sha !== sha) {
            const filePath = path.join(mainDir(side), cfgPath)

            const fileContent = await fetch(namagomiDataFileUrlBase(branch, cfgPath))
            switch (fileContent.status) {
                case 200:
                    await pipeline(await fileContent.text(),
                        createWriteStream(filePath))
                        .then(() => {
                            console.log('downloaded: ' + cfgPath)
                            if (findIndex === -1)
                                cacheJson.data.push({name: cfgPath, sha: sha})
                            else
                                cacheJson.data[findIndex].sha = sha
                        }).catch(err => {
                            console.error(err)
                            console.log('failed: ' + cfgPath + ' ' + namagomiFileUrlBase(branch, cfgPath))
                        })
                    break;
                default:
                    console.error('failed: ' + cfgPath + ' ' + namagomiFileUrlBase(branch, cfgPath))
                    console.error(`status: ${fileContent.status}`)
            }
        }
    }))
    fs.writeFileSync(namagomiCache(side), JSON.stringify(cacheJson))

    deleteFiles(dataTree, side)
}

function createEmptyJson(path: string) {
    fs.writeFileSync(path, JSON.stringify({data: [], mods: ''}))
}

function deleteFiles(dataTree: GitTree, side: string) {
    if (!fs.existsSync(namagomiCache(side))) createEmptyJson(namagomiCache(side))
    const cacheJson = JSON.parse(fs.readFileSync(namagomiCache(side), 'utf8')) as NamagomiCache

    const newCacheJson = {data: [], mods: cacheJson.mods} as NamagomiCache

    cacheJson.data.map((data) => {
        if (!dataTree.exists(data.name)) {
            const filePath = path.join(mainDir(side), data.name)
            if (fs.existsSync(filePath)) {
                console.log(`delete: ${filePath}`)
                fs.rmSync(filePath)
            }
        } else {
            newCacheJson.data.push(data)
        }
    })

    fs.writeFileSync(namagomiCache(side), JSON.stringify(newCacheJson))
}

File: src/main/scala/com/github/namagomi/main/data/NamagomiData.scala
package com.github.namagomi.main.data

import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.client.RequestBuilding.Get
import akka.http.scaladsl.unmarshalling.Unmarshal
import com.github.namagomi.main.Config
import com.github.namagomi.main.LocalPaths._
import com.github.namagomi.main.data.NamagomiCacheProtocol._
import com.github.namagomi.main.github.GitTree
import spray.json._

import java.io.PrintWriter
import java.nio.file.Paths
import scala.concurrent.duration.Duration
import scala.concurrent.{Await, ExecutionContextExecutor}
import scala.io.Source
import scala.reflect.io.File

object NamagomiData {
  implicit val system: ActorSystem = ActorSystem()
  implicit val executionContext: ExecutionContextExecutor = system.dispatcher

  def downloadAllDataFiles(branch: String)(side: String): Unit = {
    val tree = new GitTree().build("NamagomiNetwork", "Namagomi-mod", branch)
    tree.getData("data") match {
      case Some(value) =>
        val dataTree = new GitTree().build("NamagomiNetwork", "Namagomi-mod", value.data.sha)
        val dataPaths = dataTree.getAllFilePaths()
        val dataSubDirs = dataTree.getAllDirPaths()

        if (!File(namagomiCache(side)).exists)
          new PrintWriter(namagomiCache(side)).write(NamagomiCache.getEmpty)

        val source = Source.fromFile(namagomiCache(side))
        val cacheJson = source.getLines().mkString.parseJson.convertTo[NamagomiCache]
        source.close()

        if (!File(mainDir(side)).exists)
          File(mainDir(side)).createDirectory()

        dataSubDirs.foreach(dir => {
          val absDir = Paths.get(mainDir(side), dir)
          if (!absDir.toFile.exists)
            File(absDir.toString).createDirectory()
        })

        val result = dataPaths.map(cfgPath => {
          def download(): Unit = {
            val filePath = Paths.get(mainDir(side), cfgPath).toString

            val response = Await.result(Http().singleRequest(Get(Config.namagomiDataFileUrlBase(branch, cfgPath))), Duration.Inf)

            response.status match {
              case 200 =>
                val fileContent = Await.result(Unmarshal(response.entity).to[String], Duration.Inf)
                val file = new PrintWriter(filePath)
                file.write(fileContent)
                file.close()
                println(s"downloaded: $cfgPath")
              case x =>
                println(s"[ERROR] $cfgPath status code $x")
            }
          }

          dataTree.getData(cfgPath) match {
            case Some(tree) =>
              cacheJson.data.find(_.name == cfgPath) match {
                case Some(data) if data.sha != tree.data.sha =>
                  download()
                  Data(cfgPath, tree.data.sha)
                case None =>
                  download()
                  Data(cfgPath, tree.data.sha)
              }
          }
        })

        val newJson = createJson(cacheJson, result)
        val cacheWriter = new PrintWriter(namagomiCache(side))
        cacheWriter.write(newJson)
        cacheWriter.close()
    }
  }

  private def createJson(before: NamagomiCache, data: Seq[Data]): String = {
    NamagomiCache(
      data,
      before.mods
    )
      .toJson
      .compactPrint
  }
}
